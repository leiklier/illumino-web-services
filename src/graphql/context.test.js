require('dotenv').config()
const User = require('../models/user')
const Device = require('../models/device')

const db = require('../../test/util/db')
const { getTokenByUser, getTokenByDevice } = require('../lib/token')

const createContext = require('./context')

beforeAll(async () => {
	await db.create()
	await db.populate(['User', 'Device'])
})

afterAll(async () => {
	await db.destroy()
})

describe('ApolloServer context', () => {
	it('should create user context when authorized as user', async () => {
		const user = await User.findOne({ email: 'user@test.com' })
		const token = getTokenByUser(user)
		const headers = {
			authorization: `Bearer ${token}`,
		}
		const req = { headers }
		const context = await createContext({ req })

		expect(typeof context.user).toBe('object')
		expect(context.user).not.toBeNull()

		expect(context.user.id).toBe(user.id)
		expect(context.user.roles).toContain('user')
		expect(context.user.isAdmin).toBeFalsy()
		expect(context.user.isRoot).toBeFalsy()

		expect({
			...context,
			userLoader: undefined,
			deviceLoader: undefined,
		}).toMatchSnapshot()
	})

	it('should create admin context when authorized as admin', async () => {
		const user = await User.findOne({ email: 'admin@test.com' })
		const token = getTokenByUser(user)
		const headers = {
			authorization: `Bearer ${token}`,
		}
		const req = { headers }
		const context = await createContext({ req })

		expect(context.user.roles).toContain('admin')
		expect(context.user.isAdmin).toBeTruthy()
	})

	it('should create device context when authorized as device', async () => {
		const device = await Device.findOne({ mac: '00:00:00:00:00:00' })
		const token = getTokenByDevice(device)
		const headers = {
			authorization: `Bearer ${token}`,
		}
		const req = { headers }
		const context = await createContext({ req })

		expect(typeof context.device).toBe('object')
		expect(context.device).not.toBeNull()
		expect(context.device.id).toBe(device.id)

		expect({
			...context,
			userLoader: undefined,
			deviceLoader: undefined,
		}).toMatchSnapshot()
	})

	it('should set isDeploying flag when valid deploy key is provided', async () => {
		const headers = {
			authorization: `Mutual ${process.env.DEPLOY_KEY}`,
		}
		const req = { headers }
		const context = await createContext({ req })
		expect(context.isDeploying).toBeTruthy()
		expect(context.user).toBeFalsy()
		expect(context.device).toBeFalsy()
	})

	it('should not set isDeploying flag when invalid deploy key is provided', async () => {
		const headers = {
			authorization: `Mutual invalid_deply_key`,
		}
		const req = { headers }
		const context = await createContext({ req })
		expect(context.isDeploying).toBeFalsy()
		expect(context.user).toBeFalsy()
		expect(context.device).toBeFalsy()
	})

	it('should not create context without auth header', async () => {
		const headers = {
			authorization: '',
		}
		const req = { headers }
		const context = await createContext({ req })
		expect(context.user).toBeFalsy()
		expect(context.device).toBeFalsy()
		expect(context.isDeploying).toBeFalsy()
		expect(context).toMatchSnapshot()
	})
})
