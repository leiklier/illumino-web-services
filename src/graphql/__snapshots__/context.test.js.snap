// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApolloServer context should create device context when authorized as device 1`] = `
Object {
  "device": Object {
    "__v": 0,
    "_id": "5cd5b96cd5186963186be748",
    "authKey": null,
    "hasOwner": true,
    "id": "5cd5b96cd5186963186be748",
    "lastSeenAt": 2019-05-10T17:48:28.614Z,
    "mac": "00:00:00:00:00:00",
    "managers": Array [],
    "name": "Jenny sin alarm",
    "owner": Object {
      "__v": 0,
      "_id": "5cd5b885badacd5a992b0c9f",
      "devicesManaging": Array [],
      "devicesOwning": Array [
        "5cd5b96cd5186963186be748",
      ],
      "email": "user@test.com",
      "firstName": "Roli Jenny",
      "id": "5cd5b885badacd5a992b0c9f",
      "isAdmin": false,
      "isRoot": false,
      "lastName": "Oritseweyimi",
      "password": "$2a$10$iOiqxZbDw3Ekl4pUpfJmouOSnSPaJmRggIWV7sTo0.hUkcTeMcxqC",
      "roles": Array [
        "user",
      ],
    },
    "pin": null,
  },
  "deviceLoader": undefined,
  "isAuth": true,
  "measurementLoader": DataLoader {
    "_batchLoadFn": [Function],
    "_options": undefined,
    "_promiseCache": Map {},
    "_queue": Array [],
  },
  "user": null,
  "userLoader": undefined,
}
`;

exports[`ApolloServer context should create user context when authorized as user 1`] = `
Object {
  "device": null,
  "deviceLoader": undefined,
  "isAuth": true,
  "measurementLoader": DataLoader {
    "_batchLoadFn": [Function],
    "_options": undefined,
    "_promiseCache": Map {},
    "_queue": Array [],
  },
  "user": Object {
    "__v": 0,
    "_id": "5cd5b885badacd5a992b0c9f",
    "devicesManaging": Array [],
    "devicesOwning": Array [
      Object {
        "__v": 0,
        "_id": "5cd5b96cd5186963186be748",
        "authKey": "$2a$10$0GlET4dxohBrI.BpRXcLhe6XnUyF3n5ElDy.JMD5/x7mAAxpvBS/u",
        "hasOwner": true,
        "id": "5cd5b96cd5186963186be748",
        "lastSeenAt": 2019-05-10T17:48:28.614Z,
        "mac": "00:00:00:00:00:00",
        "managers": Array [],
        "name": "Jenny sin alarm",
        "owner": "5cd5b885badacd5a992b0c9f",
      },
    ],
    "email": "user@test.com",
    "firstName": "Roli Jenny",
    "id": "5cd5b885badacd5a992b0c9f",
    "isAdmin": false,
    "isRoot": false,
    "lastName": "Oritseweyimi",
    "password": null,
    "roles": Array [
      "user",
    ],
  },
  "userLoader": undefined,
}
`;

exports[`ApolloServer context should not create context without auth header 1`] = `
Object {
  "deviceLoader": DataLoader {
    "_batchLoadFn": [Function],
    "_options": undefined,
    "_promiseCache": Map {},
    "_queue": Array [],
  },
  "measurementLoader": DataLoader {
    "_batchLoadFn": [Function],
    "_options": undefined,
    "_promiseCache": Map {},
    "_queue": Array [],
  },
  "userLoader": DataLoader {
    "_batchLoadFn": [Function],
    "_options": undefined,
    "_promiseCache": Map {},
    "_queue": Array [],
  },
}
`;
